src/zenml/utils/metadata_utils.py

```diff
@@ -0,0 +1,339 @@
+            resource_id=mv.id,
+            resource_type=MetadataResourceTypes.MODEL_VERSION,
+        )
+
+    # If the user provides a model version id, we use the client to fetch it and
+    # attach the metadata to it.
+    elif model_version_id is not None:
+        model_version_id = client.get_model_version(
+            model_version_name_or_number_or_id=model_version_id
+        ).id
+        client.create_run_metadata(
+            metadata=metadata,
+            resource_id=model_version_id,
+            resource_type=MetadataResourceTypes.MODEL_VERSION,
+        )
+
+    # If the user provides an artifact name, there are three possibilities. If
+    # an artifact version is also provided with the name, we use both to fetch
+    # the artifact version and use it to log the metadata. If no version is
+    # provided, if the function is called within a step we search the artifacts
+    # of the step if not we fetch the latest version and attach the metadata
+    # to the latest version.
+    elif artifact_name is not None:
+        if artifact_version:
+            artifact_version_model = client.get_artifact_version(
+                name_id_or_prefix=artifact_name, version=artifact_version
+            )
+            client.create_run_metadata(
+                metadata=metadata,
+                resource_id=artifact_version_model.id,
+                resource_type=MetadataResourceTypes.ARTIFACT_VERSION,
+            )
+        else:
+            step_context = None
+            with contextlib.suppress(RuntimeError):
+                step_context = get_step_context()
+
+            if step_context:
+                step_context.add_output_metadata(
+                    metadata=metadata, output_name=artifact_name
+                )
+            else:
+                artifact_version_model = client.get_artifact_version(
+                    name_id_or_prefix=artifact_name
+                )
+                client.create_run_metadata(
+                    metadata=metadata,
+                    resource_id=artifact_version_model.id,
+                    resource_type=MetadataResourceTypes.ARTIFACT_VERSION,
+                )
+
+    # If the user directly provides an artifact_version_id, we use the client to
+    # fetch is and attach the metadata accordingly.
+    elif artifact_version_id is not None:
+        artifact_version_model = client.get_artifact_version(
+            name_id_or_prefix=artifact_version_id,
+        )
+        client.create_run_metadata(
+            metadata=metadata,
+            resource_id=artifact_version_model.id,
+            resource_type=MetadataResourceTypes.ARTIFACT_VERSION,
+        )
+
+    # If every additional value is None, that means we are calling it bare bones
+    # and this call needs to happen during a step execution. We will use the
+    # step context to fetch the step, run and possibly the model version and
+    # attach the metadata accordingly.
+    elif all(
```

Comment:
I'm not sure that I like the current logic: If this is being called from a step and the user provides an artifact name, don't you think we should also link it to the pipeline run, step run and model version?

Reply:
Yes 100% agreed

Reply:
This will be solved in a follow-up PR.